// Portod daemon listens on /run/portod.socket unix socket.
// Protocol: Varint length, TContainerRequest req | TContainerResponse rsp.

package rpc;

// List of error codes
enum EError {
    // Request executed successfully. No errors occured.
    Success = 0;
    // Unknown critical error.
    Unknown = 1;
    // Unknown or bad request
    InvalidMethod = 2;
    // Container with specified name already exists. Can't create a new one.
    ContainerAlreadyExists = 3;
    // Container with specified name doesn't exist.
    ContainerDoesNotExist = 4;
    // Unknown property specified.
    InvalidProperty = 5;
    // Unknown data specified.
    InvalidData = 6;
    // Invalid value of property or data.
    InvalidValue = 7;
    // Can't perform specified operation in current container state.
    InvalidState = 8;
    // Operation is not supported due to some reasons (old kernel version, missing
    // kernel modules, etc).
    NotSupported = 9;
    // Requested resources are not available (e.g. can't allocate requested memory
    // without overcommiting).
    ResourceNotAvailable = 10;
    // Insufficient rights for performing requested operation.
    Permission = 11;

    // Can't create new volume with specified name, because there is already one.
    VolumeAlreadyExists = 12;
    // Volume with specified name doesn't exist.
    VolumeNotFound = 13;
    // Operation failed because of insufficient disk space.
    NoSpace = 14;
    // Operation can't be performed, because container or volume are in use by
    // other simultaneous operation.
    Busy = 15;
    // Volume already linked with container
    VolumeAlreadyLinked = 16;
    // Volume not linked with container
    VolumeNotLinked = 17;

    LayerAlreadyExists = 18;
    LayerNotFound = 19;

    NoValue = 20;

    VolumeNotReady = 21;

    // Cannot parse or execute command
    InvalidCommand = 22;

    // Error code is lost or came from future
    LostError = 23;

    // Reserved error code used by Porto internals. Can't be returned to a user.
    Queued = 1000;
}

message TContainerRequest {
    optional TContainerCreateRequest create = 1;
    optional TContainerDestroyRequest destroy = 2;
    optional TContainerListRequest list = 3;
    optional TContainerGetPropertyRequest getProperty = 4;
    optional TContainerSetPropertyRequest setProperty = 5;
    optional TContainerGetDataRequest getData = 6;

    optional TContainerStartRequest start = 7;
    optional TContainerStopRequest stop = 8;
    optional TContainerPauseRequest pause = 9;
    optional TContainerResumeRequest resume = 10;
    optional TContainerPropertyListRequest propertyList = 11;
    optional TContainerDataListRequest dataList = 12; // deprecated
    optional TContainerKillRequest kill = 13;
    optional TVersionRequest version = 14;
    optional TContainerGetRequest get = 15;
    optional TContainerWaitRequest wait = 16;
    optional TContainerCreateRequest createWeak = 17;
    optional TContainerRespawnRequest Respawn = 18;

    optional TVolumePropertyListRequest listVolumeProperties = 103;
    optional TVolumeCreateRequest createVolume = 104;
    optional TVolumeLinkRequest linkVolume = 105;
    optional TVolumeLinkRequest LinkVolumeTarget = 120;
    optional TVolumeUnlinkRequest unlinkVolume = 106;
    optional TVolumeUnlinkRequest UnlinkVolumeTarget = 121;
    optional TVolumeListRequest listVolumes = 107;
    optional TVolumeTuneRequest tuneVolume = 108;

    optional TLayerImportRequest importLayer = 110;
    optional TLayerRemoveRequest removeLayer = 111;
    optional TLayerListRequest listLayers = 112;
    optional TLayerExportRequest exportLayer = 113;

    optional TLayerGetPrivateRequest getlayerprivate = 114;
    optional TLayerSetPrivateRequest setlayerprivate = 115;

    optional TStorageListRequest listStorage = 116;
    optional TStorageRemoveRequest removeStorage = 117;
    optional TStorageImportRequest importStorage = 118;
    optional TStorageExportRequest exportStorage = 119;

    optional TMetaStorage CreateMetaStorage = 122;
    optional TMetaStorage ResizeMetaStorage = 123;
    optional TMetaStorage RemoveMetaStorage = 124;

    optional TSetSymlinkRequest SetSymlink = 125;

    optional TConvertPathRequest convertPath = 200;
    optional TAttachProcessRequest attachProcess = 201;
    optional TLocateProcessRequest locateProcess = 202;

    optional TGetSystemRequest GetSystem = 300;
    optional TSetSystemRequest SetSystem = 301;
}

message TContainerResponse {
    optional EError error = 1 [ default = LostError ];
    optional string errorMsg = 2;

    optional TContainerListResponse list = 3;
    optional TContainerGetPropertyResponse getProperty = 4;
    optional TContainerGetDataResponse getData = 5;
    optional TContainerPropertyListResponse propertyList = 6;
    optional TContainerDataListResponse dataList = 7; // deprecated
    optional TVersionResponse version = 8;

    optional TVolumeListResponse volumeList = 9;
    optional TContainerGetResponse get = 10;
    optional TContainerWaitResponse wait = 11;
    optional TVolumePropertyListResponse volumePropertyList = 12;
    optional TVolumeDescription volume = 13;
    optional TLayerListResponse layers = 14;
    optional TConvertPathResponse convertPath = 15;
    optional TLayerGetPrivateResponse layer_private = 16;
    optional TStorageListResponse storageList = 17;
    optional TLocateProcessResponse locateProcess = 18;

    optional TGetSystemResponse GetSystem = 300;
    optional TSetSystemResponse SetSystem = 301;
}

message TGetSystemRequest {
}

message TGetSystemResponse {
    required string porto_version = 1;
    required string porto_revision = 2;
    required string kernel_version = 3;

    required fixed64 errors = 4;
    required fixed64 warnings = 5;
    required fixed64 porto_starts = 6;
    required fixed64 porto_uptime = 7;
    required fixed64 master_uptime = 8;

    required bool verbose = 100;
    required bool debug = 101;
    required fixed64 log_lines = 102;
    required fixed64 log_bytes = 103;

    required fixed64 stream_rotate_bytes = 104;
    required fixed64 stream_rotate_errors = 105;

    required fixed64 container_count = 200;
    required fixed64 container_limit = 201;
    required fixed64 container_running = 202;
    required fixed64 container_created = 203;
    required fixed64 container_started = 204;
    required fixed64 container_start_failed = 205;
    required fixed64 container_oom = 206;
    required fixed64 container_buried = 207;
    required fixed64 container_lost = 208;

    required fixed64 volume_count = 300;
    required fixed64 volume_limit = 301;
    required fixed64 volume_created = 303;
    required fixed64 volume_failed = 304;
    required fixed64 volume_links = 305;
    required fixed64 volume_links_mounted = 306;
    required fixed64 volume_lost = 307;

    required fixed64 layer_import = 390;
    required fixed64 layer_export = 391;
    required fixed64 layer_remove = 392;

    required fixed64 client_count = 400;
    required fixed64 client_max = 401;
    required fixed64 client_connected = 402;

    required fixed64 request_queued = 500;
    required fixed64 request_completed = 501;
    required fixed64 request_failed = 502;
    required fixed64 request_threads = 503;
    required fixed64 request_longer_1s = 504;
    required fixed64 request_longer_3s = 505;
    required fixed64 request_longer_30s = 506;
    required fixed64 request_longer_5m = 507;

    required fixed64 fail_system = 600;
    required fixed64 fail_invalid_value = 601;
    required fixed64 fail_invalid_command = 602;

    optional fixed64 network_count = 700;
}

message TSetSystemRequest {
    optional bool verbose = 100;
    optional bool debug = 101;
}

message TSetSystemResponse {
}

message TContainerCreateRequest {
    required string name = 1;
}

message TContainerDestroyRequest {
    required string name = 1;
}

message TContainerListRequest {
    optional string mask = 1;
}

message TContainerGetPropertyRequest {
    required string name = 1;
    required string property = 2;
    // update cached counters
    optional bool sync = 3;
    optional bool real = 4;
}

message TContainerSetPropertyRequest {
    required string name = 1;
    required string property = 2;
    required string value = 3;
}

message TContainerGetDataRequest {
    required string name = 1;
    required string data = 2;
    // update cached counters
    optional bool sync = 3;
    optional bool real = 4;
}

message TContainerStartRequest {
    required string name = 1;
}

message TContainerRespawnRequest {
    required string name = 1;
}

message TContainerStopRequest {
    required string name = 1;
    // Timeout in 1/1000 seconds between SIGTERM and SIGKILL, default 30s
    optional uint32 timeout_ms = 2;
}

message TContainerPauseRequest {
    required string name = 1;
}

message TContainerResumeRequest {
    required string name = 1;
}

message TContainerPropertyListRequest {
}

message TContainerDataListRequest {
}

message TContainerKillRequest {
    required string name = 1;
    required int32 sig = 2;
}

// Get Porto version
message TVersionRequest {
}

message TConvertPathRequest {
    required string path = 1;
    required string source = 2;
    required string destination = 3;
}

// Get multiple properties/data of many containers with one request
// (useful for monitoring)
message TContainerGetRequest {
    // list of containers
    repeated string name = 1;
    // list of properties/data
    repeated string variable = 2;
    // do not wait busy containers
    optional bool nonblock = 3;
    // update cached counters
    optional bool sync = 4;
    optional bool real = 5;
}

// Wait while container(s) is/are in running state
message TContainerWaitRequest {
    // list of containers
    repeated string name = 1;
    // timeout in 1/1000 seconds
    optional uint32 timeout_ms = 2;
}

// Move process into container
message TAttachProcessRequest {
    required string name = 1;
    required uint32 pid = 2;
    required string comm = 3; /* ignored if empty */
}

// Determine container by pid
message TLocateProcessRequest {
    required uint32 pid = 1;
    required string comm = 2; /* ignored if empty */
}

message TContainerListResponse {
    repeated string name = 1;
}

message TContainerGetPropertyResponse {
    required string value = 1;
}

message TContainerGetDataResponse {
    required string value = 1;
}

message TContainerPropertyListResponse {
    message TContainerPropertyListEntry {
        required string name = 1;
        required string desc = 2;
        optional bool read_only = 3;
        optional bool dynamic = 4;
    }
    repeated TContainerPropertyListEntry list = 1;
}

// deprecated
message TContainerDataListResponse {
    message TContainerDataListEntry {
        required string name = 1;
        required string desc = 2;
    }
    repeated TContainerDataListEntry list = 1;
}

message TVersionResponse {
    required string tag = 1;
    required string revision = 2;
}

message TContainerGetResponse {
    message TContainerGetValueResponse {
        required string variable = 1;
        optional EError error = 2;
        optional string errorMsg = 3;
        optional string value = 4;
    }
    message TContainerGetListResponse {
        required string name = 1;
        repeated TContainerGetValueResponse keyval = 2;
    }

    repeated TContainerGetListResponse list = 1;
}

message TContainerWaitResponse {
    required string name = 1;
}

message TConvertPathResponse {
    required string path = 1;
}

message TLocateProcessResponse {
    required string name = 1;
}

// VolumeAPI

message TVolumeProperty {
    required string name = 1;
    required string value = 2;
}

message TVolumePropertyDescription {
    required string name = 1;
    required string desc = 2;
}

message TVolumeLink {
    required string container = 1;
    optional string target = 2;
    optional bool required = 3;
    optional bool read_only = 4;
}

message TVolumeDescription {
    required string path = 1;
    repeated TVolumeProperty properties = 2;
    repeated string containers = 3;
    repeated TVolumeLink links = 4;
}

message TVolumePropertyListRequest {
}

message TVolumePropertyListResponse {
    repeated TVolumePropertyDescription properties = 1;
}

message TVolumeCreateRequest {
    optional string path = 1;
    repeated TVolumeProperty properties = 2;
}

message TVolumeLinkRequest {
    required string path = 1;
    optional string container = 2;      // default - self
    optional string target = 3;         // path in container, "" - anon
    optional bool required = 4;         // stop container at fail
    optional bool read_only = 5;
}

message TVolumeUnlinkRequest {
    required string path = 1;
    optional string container = 2;      // default - self, "***" - all
    optional bool strict = 3;           // non-lazy umount
    optional string target = 4;         // path in container, "" - anon, default - "***" - all
}

message TVolumeListRequest {
    optional string path = 1;
    optional string container = 2;
}

message TVolumeTuneRequest {
    required string path = 1;
    repeated TVolumeProperty properties = 2;
}

message TVolumeListResponse {
    repeated TVolumeDescription volumes = 1;
}

message TLayerImportRequest {
    required string layer = 1;
    required string tarball = 2;
    required bool merge = 3;
    optional string place = 4;
    optional string private_value = 5;
    optional string compress = 6;
}

message TLayerExportRequest {
    required string volume = 1;
    required string tarball = 2;
    optional string layer = 3;
    optional string place = 4;
    optional string compress = 5;
}

message TLayerRemoveRequest {
    required string layer = 1;
    optional string place = 2;
}

message TLayerListRequest {
    optional string place = 1;
    optional string mask = 2;
}

message TLayerGetPrivateRequest {
    required string layer = 1;
    optional string place = 2;
}

message TLayerSetPrivateRequest {
    required string layer = 1;
    optional string place = 2;
    required string private_value = 3;
}

message TLayerDescription {
    required string name = 1;
    required string owner_user = 2;
    required string owner_group = 3;
    required uint64 last_usage = 4;
    required string private_value = 5;
}

message TLayerListResponse {
    repeated string layer = 1;
    repeated TLayerDescription layers = 2;
}

message TLayerGetPrivateResponse {
    optional string private_value = 1;
}

message TStorageDescription {
    required string name = 1;
    required string owner_user = 2;
    required string owner_group = 3;
    required uint64 last_usage = 4;
    required string private_value = 5;
}

message TStorageListRequest {
    optional string place = 1;
    optional string mask = 2;
}

message TStorageListResponse {
    repeated TStorageDescription storages = 1;
    repeated TMetaStorage meta_storages = 2;
}

message TStorageRemoveRequest {
    required string name = 1;
    optional string place = 2;
}

message TStorageImportRequest {
    required string name = 1;
    required string tarball = 2;
    optional string place = 3;
    optional string private_value = 5;
    optional string compress = 6;
}

message TStorageExportRequest {
    required string name = 1;
    required string tarball = 2;
    optional string place = 3;
    optional string compress = 4;
}

message TMetaStorage {
    required string name = 1;
    optional string place = 2;
    optional string private_value = 3;
    optional uint64 space_limit = 4;
    optional uint64 inode_limit = 5;

    optional uint64 space_used = 6;
    optional uint64 space_available = 7;
    optional uint64 inode_used = 8;
    optional uint64 inode_available = 9;
    optional string owner_user = 10;
    optional string owner_group = 11;
    optional uint64 last_usage = 12;
}

message TSetSymlinkRequest {
    required string container = 1;
    required string symlink = 2;
    optional string target = 3;
}
